### PR TITLE
Improve authorization logic considering accessed organization resource

### PR DESCRIPTION
**Pull request by sadilchamishka**

### CODE REVIEW

**Comment by sadilchamishka** [comment]
Related Issue
- https://github.com/wso2-enterprise/identity-org-mgt-connector/issues/30

File: components/org.wso2.carbon.identity.organization.management.authz.service/src/main/java/org/wso2/carbon/identity/organization/management/authz/service/handler/OrganizationManagementAuthzHandler.java [hunk 1]
```diff
@@ -179,4 +189,14 @@ private void validateScopes(String[] tokenScopes, AuthorizationContext authoriza
             }
         }
     }
+
+    private String extractOrganizationId(String requestPath) {
+
+        if (Pattern.matches(REGEX_FOR_URLS_WITH_ORG_ID, requestPath)) {
+            String[] requestUriParts = requestPath.split(URI_SPLITTER);
```
Comment (wenodpathirana) [inline components/org.wso2.carbon.identity.organization.management.authz.service/src/main/java/org/wso2/carbon/identity/organization/management/authz/service/handler/OrganizationManagementAuthzHandler.java:196]: You can use Matcher to extract the value.

Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.organization.management.authz.service/src/main/java/org/wso2/carbon/identity/organization/management/authz/service/handler/OrganizationManagementAuthzHandler.java]: The authorization logic was changed to apply a validation when organization role access
https://github.com/wso2-extensions/identity-organization-management/pull/143/commits/7fe3192d344c1a4991dcc49063e35c603463f4a6

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java [hunk 2]
```diff
@@ -442,6 +446,16 @@ private void appendFragmentAppProperties(ServiceProvider serviceProvider) {
         serviceProvider.setSpProperties(spProperties);
     }
 
+    private void validateOrganizationApplicationsAllowedToAccess(String organizationId)
+            throws OrganizationManagementException {
+
+        String authorizedOrganizationId = getOrganizationId(); // The organization the user is authorized to access
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java:452]:
```suggestion
        String authorizedOrganizationId = getOrganizationId(); // The organization that the user is authorized to access
```

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java [hunk 3]
```diff
@@ -442,6 +446,16 @@ private void appendFragmentAppProperties(ServiceProvider serviceProvider) {
...
+        if (!StringUtils.equals(organizationId, authorizedOrganizationId)) {
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java:453]: StringUtils.equals(null, null) return true. So can we guarantee that either one or both organizationId, authorizedOrganizationId are not null?

File: components/org.wso2.carbon.identity.organization.management.role.management.service/src/main/java/org/wso2/carbon/identity/organization/management/role/management/service/RoleManagerImpl.java [hunk 4]
```diff
@@ -429,4 +439,13 @@ private Cursor decodeCursor(String cursorString) throws OrganizationManagementEx
             throw handleClientException(ERROR_CODE_ROLE_LIST_INVALID_CURSOR, cursorString);
         }
     }
+
+    private void validateOrganizationRoleAllowedToAccess(String organizationId) throws OrganizationManagementException {
+
+        String authorizedOrganizationId = getOrganizationId(); // The organization the user is authorized to access
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.organization.management.role.management.service/src/main/java/org/wso2/carbon/identity/organization/management/role/management/service/RoleManagerImpl.java:445]:
```suggestion
        String authorizedOrganizationId = getOrganizationId(); // The organization that the user is authorized to access
```

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java (hunk 3 above)
Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java:453]: `getOrganizationId()` method return the `organizationId` set in the carbon context. It is set only for organization qualified urls and organization management endpoints are exposed for organization qualified paths only. Therefore `authorizedOrganizationId` always has a valid org id.

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java (hunk 2 above)
Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java]: Fixed by - https://github.com/wso2-extensions/identity-organization-management/pull/143/commits/963ba2a414328358b466db65264da1daadf3e1eb

File: components/org.wso2.carbon.identity.organization.management.role.management.service/src/main/java/org/wso2/carbon/identity/organization/management/role/management/service/RoleManagerImpl.java (hunk 4 above)
Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.organization.management.role.management.service/src/main/java/org/wso2/carbon/identity/organization/management/role/management/service/RoleManagerImpl.java]: Fixed by - https://github.com/wso2-extensions/identity-organization-management/pull/143/commits/963ba2a414328358b466db65264da1daadf3e1eb

**Review by AnuradhaSK — APPROVED** [review]

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java (hunk 3 above)
Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java:453]: `getOrganizationId() ` method returns null when the organization management APIs called from the super organization. (The organization qualified urls no need for request from super orgs ). Therefore the above mentioned issue need to be handled properly.

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java [hunk 5]
```diff
@@ -442,6 +446,16 @@ private void appendFragmentAppProperties(ServiceProvider serviceProvider) {
         serviceProvider.setSpProperties(spProperties);
     }
 
+    private void validateOrganizationApplicationsAllowedToAccess(String organizationId)
+            throws OrganizationManagementException {
+
+        String authorizedOrganizationId = getOrganizationId(); // The organization that the user is authorized to access
```
Comment (sahandilshan) [inline components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java:452]: Formatting error: Missing the full stop from the comment.

File: components/org.wso2.carbon.identity.organization.management.role.management.service/src/main/java/org/wso2/carbon/identity/organization/management/role/management/service/RoleManagerImpl.java [hunk 6]
```diff
@@ -429,4 +439,13 @@ private Cursor decodeCursor(String cursorString) throws OrganizationManagementEx
             throw handleClientException(ERROR_CODE_ROLE_LIST_INVALID_CURSOR, cursorString);
         }
     }
+
+    private void validateOrganizationRoleAllowedToAccess(String organizationId) throws OrganizationManagementException {
+
+        String authorizedOrganizationId = getOrganizationId(); // The organization that the user is authorized to access
```
Comment (sahandilshan) [inline components/org.wso2.carbon.identity.organization.management.role.management.service/src/main/java/org/wso2/carbon/identity/organization/management/role/management/service/RoleManagerImpl.java:445]: Formatting error: Missing the full stop from the comment.

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java [hunk 7]
```diff
@@ -850,4 +856,12 @@ private TenantMgtService getTenantMgtService() {
 
         return OrganizationManagementDataHolder.getInstance().getTenantMgtService();
     }
+
+    private void validateOrganizationAllowedToAccess(String organizationId) throws OrganizationManagementException {
+
+        String authorizedOrganizationId = getOrganizationId(); // The organization that the user is authorized to access
```
Comment (sahandilshan) [inline components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java:862]: Formatting error: Missing the full stop from the comment.

File: components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java (hunk 5 above)
Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.organization.management.application/src/main/java/org/wso2/carbon/identity/organization/management/application/OrgApplicationManagerImpl.java:452]: This code block is removed after a later PR.

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java (hunk 7 above)
Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/OrganizationManagerImpl.java:862]: This code block is removed after a later PR. https://github.com/wso2-extensions/identity-organization-management-core/commit/361f104571abd702f2e3ae83e4568a1d4dd474ad#diff-e9e687f238a9891ce911066c864c216575cd81e7fc9e559efe2b4a7cc99a39eaL861

File: components/org.wso2.carbon.identity.organization.management.role.management.service/src/main/java/org/wso2/carbon/identity/organization/management/role/management/service/RoleManagerImpl.java (hunk 6 above)
Comment (sadilchamishka) [reply components/org.wso2.carbon.identity.organization.management.role.management.service/src/main/java/org/wso2/carbon/identity/organization/management/role/management/service/RoleManagerImpl.java:445]: Fixed by - https://github.com/wso2-extensions/identity-organization-management/pull/164/commits/e352b148a75d1964008c0efd6d0c20c665b07e5e

**Pull request merged**